File: hw1/merge-sort-func.js
module.exports = function mergeSort(arr) {
    if (arr.length < 2) {
        return arr;
    }

    const middle = Math.floor(arr.length / 2);
    const left = arr.slice(0, middle);
    const right = arr.slice(middle, arr.length);
    const sortedLeft = mergeSort(left);
    const sortedRight = mergeSort(right);
    return merge(sortedLeft, sortedRight);
};

function merge(left, right) {
    let results = [];

    while (left.length && right.length) {
        if (left[0] < right[0]) {
            shift(left);
        } else {
            shift(right);
        }
    }

    while (left.length) {
        shift(left);
    }

    while (right.length) {
        shift(right);
    }

    function shift(arr) {
        results.push(arr.shift());
    }

    return results;
}

File: hw2/merge_sort_func.py
import math

def merge_sort(arr):
    """

    :type arr: list
    """
    length = len(arr)

    if length < 2:
        return arr

    middle = int(math.floor(length / 2))
    left = arr[0:middle]
    right = arr[middle:]
    sorted_left = merge_sort(left)
    sorted_right = merge_sort(right)
    return merge(sorted_left, sorted_right)


def merge(left, right):
    results = []

    while len(left) and len(right):
        if left[0] < right[0]:
            results.append(left.pop(0))
        else:
            results.append(right.pop(0))

    while len(left):
        results.append(left.pop(0))

    while len(right):
        results.append(right.pop(0))

    return results